src/cargo/ops/cargo_output_metadata.rs

```diff
@@ -0,0 +1,146 @@
+    StdOut,
+}
+
+impl Decodable for OutputTo {
+    fn decode<D: Decoder>(d: &mut D) -> Result<OutputTo, D::Error> {
+        d.read_option(|d, b| {
+            if b {
+                let path = PathBuf::from(try!(d.read_str()));
+                Ok(OutputTo::Path(path))
+            } else {
+                Ok(OutputTo::StdOut)
+            }
+        })
+    }
+}
+
+pub struct OutputMetadataOptions<'a> {
+    pub features: Vec<String>,
+    pub output_format: OutputFormat,
+    pub output_to: OutputTo,
+    pub manifest_path: &'a Path,
+    pub no_default_features: bool,
+}
+
+/// Loads the manifest, resolves the dependencies of the project to the concrete
+/// used versions - considering overrides - and writes all dependencies in a TOML
+/// format to stdout or the specified file.
+///
+/// The TOML format is e.g.:
+/// ```toml
+/// root = "libA"
+///
+/// [packages.libA]
+/// dependencies = ["libB"]
+/// path = "/home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/libA-0.1"
+/// version = "0.1"
+///
+/// [packages.libB]
+/// dependencies = []
+/// path = "/home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/libB-0.4"
+/// version = "0.4"
+///
+/// [packages.libB.features]
+/// featureA = ["featureB"]
+/// featureB = []
+/// ```
+pub fn output_metadata(opt: OutputMetadataOptions, config: &Config) -> CargoResult<()> {
+    let deps = try!(resolve_dependencies(opt.manifest_path,
+                                         config,
+                                         opt.features,
+                                         opt.no_default_features));
+    let (resolved_deps, packages) = deps;
+
+    #[derive(RustcEncodable)]
+    struct RootPackageInfo<'a> {
+        name: &'a str,
+        version: String,
+        features: Option<&'a HashMap<String, Vec<String>>>,
+    }
+
+    #[derive(RustcEncodable)]
+    struct ExportInfo<'a> {
+        root: RootPackageInfo<'a>,
+        packages: Vec<&'a Package>,
+    }
+
+    let mut output = ExportInfo {
+        root: RootPackageInfo {
+            name: resolved_deps.root().name(),
+            version: format!("{}", resolved_deps.root().version()),
+            features: None,
+        },
+        packages: Vec::new(),
+    };
+
+    for package in packages.iter() {
+        output.packages.push(&package);
+        if package.package_id() == resolved_deps.root() {
+            let features = package.manifest().summary().features();
+            if !features.is_empty() {
+                output.root.features = Some(features);
+            }
+        }
+    }
+
+    let serialized_str = match opt.output_format {
+        OutputFormat::Toml => toml::encode_str(&output),
+        OutputFormat::Json => try!(json::encode(&output)),
+    };
+
+    match opt.output_to {
+        OutputTo::StdOut => println!("{}", serialized_str),
+        OutputTo::Path(ref path) => {
+            let mut file = try!(fs::File::create(path));
+            try!(file.write_all(serialized_str.as_bytes()));
```

Comment:
There's a helper function `paths::write` which should encompass this write-the-whole-file logic. This may also want to chain errors to give some more contextual information about what failed.